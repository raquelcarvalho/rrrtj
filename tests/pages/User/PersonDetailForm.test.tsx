//Aqui, verificamos o tipo, o espaço reservado, o rótulo dos campos de entrada, bem como o tipo e o texto do botão.

import React from "react";
import { PersonalDetailForm } from "../../../src/pages/User/PersonalDetailForm";
import ReduxPersonalDetailForm from "../../../src/pages/User/PersonalDetailForm";
import { Provider } from "react-redux";
import { combineReducers, createStore } from "redux";
import { reducer as formReducer } from "redux-form";
import { ShallowWrapper } from "enzyme";
import { shallowToJson } from "enzyme-to-json";
import { mount, shallow } from "../../setup/test-setup";

const mockfn = jest.fn();

describe("<PersonalDetailForm />", () => {
  let wrapper: any;
  let store: any;

  const props: any = {
    handleSubmit: mockfn,
  };

  beforeEach(() => {
    wrapper = shallow(<PersonalDetailForm {...props} />);
  });

  it("should match the snapshot", () => {
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  describe("defines user form fields", () => {
    it("renders user name field", () => {
      const userNameField = wrapper.find('[name="userName"]');

      expect(userNameField.prop("type")).toBe("text");
      expect(userNameField.prop("label")).toBe("Name *");
      expect(userNameField.prop("placeHolder")).toBe("Enter User Name");
    });

    it("renders user dob field", () => {
      const userDOBField = wrapper.find('[name="userDOB"]');

      expect(userDOBField.prop("type")).toBe("date");
      expect(userDOBField.prop("label")).toBe("DOB *");
      expect(userDOBField.prop("placeHolder")).toBe("Enter Date of Birth");
    });

    it("renders user age field", () => {
      const userAgeField = wrapper.find('[name="userAge"]');

      expect(userAgeField.prop("type")).toBe("text");
      expect(userAgeField.prop("label")).toBe("Age *");
      expect(userAgeField.prop("placeHolder")).toBe("Enter User Age");
    });

    it("renders gender field", () => {
      const genderField = wrapper.find('[name="userGender"]');

      expect(genderField.prop("type")).toBe("text");
      expect(genderField.prop("label")).toBe("Gender *");
      expect(genderField.prop("placeHolder")).toBe("Select Gender");
    });

    it("renders mobile number field", () => {
      const mobileNumberField = wrapper.find('[name="userMobileNumber"]');

      expect(mobileNumberField.prop("type")).toBe("text");
      expect(mobileNumberField.prop("label")).toBe("Mobile No *");
      expect(mobileNumberField.prop("placeHolder")).toBe("Enter Mobile Number");
    });

    it("renders address field", () => {
      const addressField = wrapper.find('[name="userAddress"]');

      expect(addressField.prop("type")).toBe("textarea");
      expect(addressField.prop("label")).toBe("Address *");
      expect(addressField.prop("placeHolder")).toBe("Enter Address");
    });

    it("renders next button", () => {
      const nextButton = wrapper.find("Button").dive();
      expect(nextButton.prop("type")).toBe("submit");
      expect(nextButton.text()).toEqual("Next <FaChevronRight />");
    });
  });

  //!simulamos cada entrada e verificamos se a parte da validação e o texto da validação funcionam ou não.
  describe("form validation", () => {
    beforeEach(() => {
      store = createStore(combineReducers({ form: formReducer }));
      wrapper = mount(
        <Provider store={store}>
          <ReduxPersonalDetailForm {...props} />
        </Provider>
      );
    });

    describe("input#username", () => {
      let nameField: ShallowWrapper;

      beforeEach(() => {
        nameField = wrapper.find('input[name="userName"]').first();
      });

      it("shows error when name is set to blank", () => {
        nameField.simulate("blur");
        const errorBlock = wrapper.find(".text-danger");

        expect(errorBlock).toHaveLength(1);
        expect(errorBlock.text()).toBe("User name required");
      });
    });

    describe("input#userDOB", () => {
      let dobField: ShallowWrapper;

      beforeEach(() => {
        dobField = wrapper.find('input[name="userDOB"]').first();
      });

      it("shows error when DOB is set to blank", () => {
        dobField.simulate("blur");
        const errorBlock = wrapper.find(".text-danger");

        expect(errorBlock).toHaveLength(1);
        expect(errorBlock.text()).toBe("DOB required");
      });
    });

    describe("input#userAge", () => {
      let ageField: ShallowWrapper;

      beforeEach(() => {
        ageField = wrapper.find('input[name="userAge"]').first();
      });

      it("shows error when age is set to blank", () => {
        ageField.simulate("blur");
        const errorBlock = wrapper.find(".text-danger");

        expect(errorBlock).toHaveLength(1);
        expect(errorBlock.text()).toBe("Age required");
      });
    });

    describe("input#userGender", () => {
      let genderField: ShallowWrapper;

      beforeEach(() => {
        genderField = wrapper.find('select[name="userGender"]').first();
      });

      it("shows error when gender is set to blank", () => {
        genderField.simulate("blur");
        const errorBlock = wrapper.find(".text-danger");

        expect(errorBlock).toHaveLength(1);
        expect(errorBlock.text()).toBe("Gender required");
      });
    });

    describe("input#userMobileNumber", () => {
      let mobileField: ShallowWrapper;

      beforeEach(() => {
        mobileField = wrapper.find('input[name="userMobileNumber"]').first();
      });

      it("shows error when mobile number is set to blank", () => {
        mobileField.simulate("blur");
        const errorBlock = wrapper.find(".text-danger");

        expect(errorBlock).toHaveLength(1);
        expect(errorBlock.text()).toBe("Mobile number required");
      });
    });

    describe("input#userAddress", () => {
      let addressField: ShallowWrapper;

      beforeEach(() => {
        addressField = wrapper.find('textarea[name="userAddress"]').first();
      });

      it("shows error when address is set to blank", () => {
        addressField.simulate("blur");
        const errorBlock = wrapper.find(".text-danger");

        expect(errorBlock).toHaveLength(1);
        expect(errorBlock.text()).toBe("Address required");
      });
    });
  });
});
