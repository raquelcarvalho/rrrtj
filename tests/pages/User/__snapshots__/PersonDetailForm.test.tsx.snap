// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PersonalDetailForm /> should match the snapshot 1`] = `
<form
  noValidate={true}
  onSubmit={[MockFunction]}
>
  <Col
    sm="12"
    tag="div"
    widths={
      Array [
        "xs",
        "sm",
        "md",
        "lg",
        "xl",
      ]
    }
  >
    <Card
      className="card-border"
      tag="div"
    >
      <CardBody
        tag="div"
      >
        <FormGroup
          row={true}
          tag="div"
        >
          <Col
            lg="12"
            tag="div"
            widths={
              Array [
                "xs",
                "sm",
                "md",
                "lg",
                "xl",
              ]
            }
            xs="12"
          >
            <Field
              component={[Function]}
              label="Name *"
              name="userName"
              normalize={[Function]}
              placeHolder="Enter User Name"
              type="text"
            />
          </Col>
        </FormGroup>
        <FormGroup
          row={true}
          tag="div"
        >
          <Col
            lg="6"
            tag="div"
            widths={
              Array [
                "xs",
                "sm",
                "md",
                "lg",
                "xl",
              ]
            }
            xs="12"
          >
            <Field
              component={[Function]}
              label="DOB *"
              maxDate="2020-05-15"
              name="userDOB"
              onChange={[Function]}
              placeHolder="Enter Date of Birth"
              type="date"
            />
          </Col>
          <Col
            lg="6"
            tag="div"
            widths={
              Array [
                "xs",
                "sm",
                "md",
                "lg",
                "xl",
              ]
            }
            xs="12"
          >
            <Field
              component={[Function]}
              label="Age *"
              name="userAge"
              placeHolder="Enter User Age"
              type="text"
            />
          </Col>
        </FormGroup>
        <FormGroup
          row={true}
          tag="div"
        >
          <Col
            lg="6"
            tag="div"
            widths={
              Array [
                "xs",
                "sm",
                "md",
                "lg",
                "xl",
              ]
            }
            xs="12"
          >
            <Field
              component={[Function]}
              datas={
                Array [
                  Object {
                    "label": "Male",
                    "value": "male",
                  },
                  Object {
                    "label": "Female",
                    "value": "female",
                  },
                  Object {
                    "label": "Other",
                    "value": "other",
                  },
                ]
              }
              label="Gender *"
              name="userGender"
              placeHolder="Select Gender"
              type="text"
            />
          </Col>
          <Col
            lg="6"
            tag="div"
            widths={
              Array [
                "xs",
                "sm",
                "md",
                "lg",
                "xl",
              ]
            }
            xs="12"
          >
            <Field
              component={[Function]}
              label="Mobile No *"
              name="userMobileNumber"
              normalize={[Function]}
              placeHolder="Enter Mobile Number"
              type="text"
            />
          </Col>
        </FormGroup>
        <FormGroup
          row={true}
          tag="div"
        >
          <Col
            lg="12"
            tag="div"
            widths={
              Array [
                "xs",
                "sm",
                "md",
                "lg",
                "xl",
              ]
            }
            xs="12"
          >
            <Field
              component={[Function]}
              label="Address *"
              name="userAddress"
              placeHolder="Enter Address"
              type="textarea"
            />
          </Col>
        </FormGroup>
      </CardBody>
      <div
        style={
          Object {
            "paddingBottom": 30,
          }
        }
      >
        <Button
          className="float-right"
          color="success"
          style={
            Object {
              "marginRight": "10px",
            }
          }
          tag="button"
          type="submit"
        >
          Next 
          <FaChevronRight
            className="button-padding"
            size={18}
          />
        </Button>
      </div>
    </Card>
  </Col>
</form>
`;
